`localPassC` pragma has no effect on nixpkgs `nim`

## Describe the bug

When compiling nim modules the `localPassC` pragma is ignored.
I think this is a by-product (or intended effect?) of the `extra-mangling-2.patch`.

This issue came up when preparing to bump `nimble` to v0.16.3 which now makes use of `zippy` as a submodule.

## Steps To Reproduce

Steps to reproduce the behavior:

```sh
git clone git@github.com:guzba/zippy && cd zippy
nix run "github:nixos/nixpkgs/nixos-unstable#nim" -- --path:src/ c examples/ziparchive_extract.nim
```

## Expected behavior

Compiles without error:

Compiler built with this [commit](https://github.com/NixOS/nixpkgs/compare/master...daylinmorgan:nixpkgs:nim-no-mangle) removing the name mangling patch is available from my cache, and compiles without error.

```sh
nix run "github:daylinmorgan/nixpkgs/nim-no-mangle#nim" --extra-substituters "https://attic.dayl.in/oizys" --extra-trusted-public-keys "oizys:DSw3mwVMM/Y+PXSVpkDlU5dLwlORuiJRGPkwr5INSMc=" -- --path:src/ c examples/ziparchive_extract.nim
```

## Screenshots

<!-- If applicable, add screenshots to help explain your problem: -->

## Additional context

Error messaging compiling zippy:

```txt
In file included from /home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fnqyre32_fvzq.nim.c:9:
/nix/store/v7dbfh5n7az2lcap0z1cv4jq0bikya8p-gcc-13.3.0/lib/gcc/x86_64-unknown-linux-gnu/13.3.0/include/tmmintrin.h: In function ‘adler32_ssse3__OOZfepZmvcclZnqyre3295fvzq_u84’:
/nix/store/v7dbfh5n7az2lcap0z1cv4jq0bikya8p-gcc-13.3.0/lib/gcc/x86_64-unknown-linux-gnu/13.3.0/include/tmmintrin.h:112:1: error: inlining failed in call to ‘always_inline’ ‘_mm_maddubs_epi16’: target specific option mismatch
  112 | _mm_maddubs_epi16 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~~
/home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fnqyre32_fvzq.nim.c:227:96: note: called from here
  227 |                            nimln_(95);                                     mad2 = _mm_maddubs_epi16(bytes2, tap2);
      |                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/nix/store/v7dbfh5n7az2lcap0z1cv4jq0bikya8p-gcc-13.3.0/lib/gcc/x86_64-unknown-linux-gnu/13.3.0/include/tmmintrin.h:112:1: error: inlining failed in call to ‘always_inline’ ‘_mm_maddubs_epi16’: target specific option mismatch
  112 | _mm_maddubs_epi16 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~~
/home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fnqyre32_fvzq.nim.c:222:96: note: called from here
  222 |                            nimln_(92);                                     mad1 = _mm_maddubs_epi16(bytes1, tap1);
      |                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/nix/store/v7dbfh5n7az2lcap0z1cv4jq0bikya8p-gcc-13.3.0/lib/gcc/x86_64-unknown-linux-gnu/13.3.0/include/tmmintrin.h:112:1: error: inlining failed in call to ‘always_inline’ ‘_mm_maddubs_epi16’: target specific option mismatch
  112 | _mm_maddubs_epi16 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~~
/home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fnqyre32_fvzq.nim.c:222:96: note: called from here
  222 |                            nimln_(92);                                     mad1 = _mm_maddubs_epi16(bytes1, tap1);
      |                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

compilation terminated due to -fmax-errors=3.
Error: execution of an external compiler program '/nix/store/hgw57m0jh2w10jsv8x86ng267pm5sla8-gcc-wrapper-13.3.0/bin/gcc -c -w -fmax-errors=3 -pthread   -I/nix/store/hw3a3n3b6yxsn3bp6j92jh0745m6sp9c-nim-unwrapped-2.2.0/nim/lib -I/home/daylin/tmp/zippy/examples -o /home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fnqyre32_fvzq.nim.c.o /home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fnqyre32_fvzq.nim.c' failed with exit code: 1


In file included from /home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fpep32_fvzq.nim.c:10:
/home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fpep32_fvzq.nim.c: In function ‘crc32_sse41_pcmul__OOZfepZmvcclZpep3295fvzq_u39’:
/home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fpep32_fvzq.nim.c:187:62: error: ‘__builtin_ia32_pclmulqdq128’ needs isa option -mpclmul -msse2
  187 |                         nimln_(68);                     x5 = _mm_clmulepi64_si128(x1, x0, ((NI)0));
      |                                                              ^~~~~~~~~~~~~~~~~~~~
/home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fpep32_fvzq.nim.c:188:62: error: ‘__builtin_ia32_pclmulqdq128’ needs isa option -mpclmul -msse2
  188 |                         nimln_(69);                     x6 = _mm_clmulepi64_si128(x2, x0, ((NI)0));
      |                                                              ^~~~~~~~~~~~~~~~~~~~
/home/daylin/.cache/nim/ziparchive_extract_d/@z..@ffep@fmvccl@fpep32_fvzq.nim.c:189:62: error: ‘__builtin_ia32_pclmulqdq128’ needs isa option -mpclmul -msse2
  189 |                         nimln_(70);                     x7 = _mm_clmulepi64_si128(x3, x0, ((NI)0));
      |                                                              ^~~~~~~~~~~~~~~~~~~~
compilation terminated due to -fmax-errors=3.
```

<!-- Add any other context about the problem here. -->

## Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->

```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.11.6, NixOS, 24.11 (Vicuna), 24.11.20241111.dc460ec`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.92.0-dev-pre20241112-4377793
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux, x86_64-v4-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/daylin/.config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/daylin/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/daylin/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/daylin/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/b1vx82bcqr6n6cafmy2hd65v0i7q53i5-lix-2.92.0-dev-pre20241112-4377793/share`
 - channels(root): `"nixos-23.11"`
 - nixpkgs: `/nix/store/4r8s42c9mwfvgdlxv0izb1cmlzrsb5nz-source`
```

## Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@ehmry
@eveeifyeve

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/ni/nim-unwrapped-1_0/extra-mangling.patch
+++ b/pkgs/by-name/ni/nim-unwrapped-1_0/extra-mangling.patch
@@ -1,13 +1,17 @@
 diff --git a/compiler/modulepaths.nim b/compiler/modulepaths.nim
-index e80ea3fa6..8ecf27a85 100644
+index fa8fab08a..63b0cb44d 100644
 --- a/compiler/modulepaths.nim
 +++ b/compiler/modulepaths.nim
-@@ -70,6 +70,13 @@ proc checkModuleName*(conf: ConfigRef; n: PNode; doLocalError=true): FileIndex =
+@@ -73,6 +73,17 @@ proc checkModuleName*(conf: ConfigRef; n: PNode; doLocalError=true): FileIndex =
    else:
      result = fileInfoIdx(conf, fullPath)
  
 +proc rot13(result: var string) =
-+  for i, c in result:
++  # don't mangle .nim
++  let finalIdx =
++    if result.endsWith(".nim"): result.len - 4
++    else: result.len
++  for i, c in result[0..<finalIdx]:
 +    case c
 +    of 'a'..'m', 'A'..'M': result[i] = char(c.uint8 + 13)
 +    of 'n'..'z', 'N'..'Z': result[i] = char(c.uint8 - 13)
@@ -16,7 +20,7 @@ index e80ea3fa6..8ecf27a85 100644
  proc mangleModuleName*(conf: ConfigRef; path: AbsoluteFile): string =
    ## Mangle a relative module path to avoid path and symbol collisions.
    ##
-@@ -78,9 +85,11 @@ proc mangleModuleName*(conf: ConfigRef; path: AbsoluteFile): string =
+@@ -81,9 +92,11 @@ proc mangleModuleName*(conf: ConfigRef; path: AbsoluteFile): string =
    ##
    ## Example:
    ## `foo-#head/../bar` becomes `@foo-@hhead@s..@sbar`
@@ -30,10 +34,10 @@ index e80ea3fa6..8ecf27a85 100644
    result = path.multiReplace({"@@": "@", "@h": "#", "@s": "/", "@m": "", "@c": ":"})
 +  rot13(result)
 diff --git a/compiler/msgs.nim b/compiler/msgs.nim
-index 3f386cc61..054f7f647 100644
+index ae4bcfcb8..1ad7e5c08 100644
 --- a/compiler/msgs.nim
 +++ b/compiler/msgs.nim
-@@ -659,8 +659,10 @@ proc uniqueModuleName*(conf: ConfigRef; fid: FileIndex): string =
+@@ -661,8 +661,10 @@ proc uniqueModuleName*(conf: ConfigRef; fid: FileIndex): string =
    for i in 0..<trunc:
      let c = rel[i]
      case c

--- a/pkgs/by-name/ni/nim-unwrapped-2_2/extra-mangling-2.patch
+++ b/pkgs/by-name/ni/nim-unwrapped-2_2/extra-mangling-2.patch
@@ -1,34 +1,3 @@
-diff --git a/compiler/modulepaths.nim b/compiler/modulepaths.nim
-index c9e6060e5..acb289498 100644
---- a/compiler/modulepaths.nim
-+++ b/compiler/modulepaths.nim
-@@ -79,6 +79,13 @@ proc checkModuleName*(conf: ConfigRef; n: PNode; doLocalError=true): FileIndex =
-   else:
-     result = fileInfoIdx(conf, fullPath)
- 
-+proc rot13(result: var string) =
-+  for i, c in result:
-+    case c
-+    of 'a'..'m', 'A'..'M': result[i] = char(c.uint8 + 13)
-+    of 'n'..'z', 'N'..'Z': result[i] = char(c.uint8 - 13)
-+    else: discard
-+
- proc mangleModuleName*(conf: ConfigRef; path: AbsoluteFile): string =
-   ## Mangle a relative module path to avoid path and symbol collisions.
-   ##
-@@ -87,9 +94,11 @@ proc mangleModuleName*(conf: ConfigRef; path: AbsoluteFile): string =
-   ##
-   ## Example:
-   ## `foo-#head/../bar` becomes `@foo-@hhead@s..@sbar`
--  "@m" & relativeTo(path, conf.projectPath).string.multiReplace(
-+  result = "@m" & relativeTo(path, conf.projectPath).string.multiReplace(
-     {$os.DirSep: "@s", $os.AltSep: "@s", "#": "@h", "@": "@@", ":": "@c"})
-+  rot13(result)
- 
- proc demangleModuleName*(path: string): string =
-   ## Demangle a relative module path.
-   result = path.multiReplace({"@@": "@", "@h": "#", "@s": "/", "@m": "", "@c": ":"})
-+  rot13(result)
 diff --git a/compiler/modulegraphs.nim b/compiler/modulegraphs.nim
 index 77762d23a..59dd8903a 100644
 --- a/compiler/modulegraphs.nim
@@ -46,3 +15,38 @@ index 77762d23a..59dd8903a 100644
        result.add c
      of {os.DirSep, os.AltSep}:
        result.add 'Z' # because it looks a bit like '/'
+diff --git a/compiler/modulepaths.nim b/compiler/modulepaths.nim
+index c9e6060e5..2b349f27c 100644
+--- a/compiler/modulepaths.nim
++++ b/compiler/modulepaths.nim
+@@ -79,6 +79,17 @@ proc checkModuleName*(conf: ConfigRef; n: PNode; doLocalError=true): FileIndex =
+   else:
+     result = fileInfoIdx(conf, fullPath)
+ 
++proc rot13(result: var string) =
++  # don't mangle .nim
++  let finalIdx =
++    if result.endsWith(".nim"): result.len - 4
++    else: result.len
++  for i, c in result[0..<finalIdx]:
++    case c
++    of 'a'..'m', 'A'..'M': result[i] = char(c.uint8 + 13)
++    of 'n'..'z', 'N'..'Z': result[i] = char(c.uint8 - 13)
++    else: discard
++
+ proc mangleModuleName*(conf: ConfigRef; path: AbsoluteFile): string =
+   ## Mangle a relative module path to avoid path and symbol collisions.
+   ##
+@@ -87,9 +98,11 @@ proc mangleModuleName*(conf: ConfigRef; path: AbsoluteFile): string =
+   ##
+   ## Example:
+   ## `foo-#head/../bar` becomes `@foo-@hhead@s..@sbar`
+-  "@m" & relativeTo(path, conf.projectPath).string.multiReplace(
++  result = "@m" & relativeTo(path, conf.projectPath).string.multiReplace(
+     {$os.DirSep: "@s", $os.AltSep: "@s", "#": "@h", "@": "@@", ":": "@c"})
++  rot13(result)
+ 
+ proc demangleModuleName*(path: string): string =
+   ## Demangle a relative module path.
+   result = path.multiReplace({"@@": "@", "@h": "#", "@s": "/", "@m": "", "@c": ":"})
++  rot13(result)

